kickoff txt registry

ideahitme opened the pull request:
Changes introduced by this PR:
1. RecordType field is part of Endpoint struct 
    a. From source initialized to "", from dnsprovider inherited from dns provider record type
    b. AWS Tests are adjusted to use NewEndpoint initialiser and now allow to inspect the type 
2. InMemory DNS Provider is changed to handle recordtype passed from endpoint struct
3. TXT Registry implementation
    a. Records() collapses ownership TXT records into Labels["owner"] field
    b. ApplyChanges() filters only owned records. For create/delete lists adds TXT records to the list 
    c. Support various formats for txt record DNSName (i.e. prefixed or 1-1 mapping) 
4. TXT registry is integrated as part of the flow (new flags are added, selectable registry) 
5. Various test improvements

ideahitme commented:
/cc @linki I would like to add some minor changes, but PR can be reviewed

ideahitme commented:
need to rebase to fix failing build

hjacobs commented on endpoint/endpoint.go:
```diff
@@ -31,15 +31,18 @@ type Endpoint struct {
 
 // NewEndpoint initialization method to be used to create an endpoint
-func NewEndpoint(dnsName, target string) *Endpoint {
+func NewEndpoint(dnsName, target, recordType string) *Endpoint {
```
Is Endpoint still the right name? It looks like it becomes a DNS record struct now..

hjacobs replied:
Maybe just rename it to fit all cases (a `TXT` record is not an "endpoint" for me)?

ideahitme replied:
yes, were considering splitting into two different structs, one for endpoints (collected from k8s resources) and the other is the struct defined for dns provider records. However, afaik we should be able to define all dns record parameters (ttl, record type etc) in the resource annotations, so for now we will just stick to a single struct

szuecs commented on endpoint/endpoint_test.go:
```diff
@@ -19,16 +19,16 @@ package endpoint
 func TestNewEndpoint(t *testing.T) {
-	e := NewEndpoint("example.org", "1.2.3.4")
-	if e.DNSName != "example.org" || e.Target != "1.2.3.4" {
+	e := NewEndpoint("example.org", "1.2.3.4", "CNAME")
```
CNAME with an IP target?
It should be another DNS entry as target

ideahitme replied:
yeah, but this is not to test reality.

szuecs commented on internal/testutils/endpoint.go:
```diff
@@ -17,14 +17,39 @@ limitations under the License.
+			return true
+		}
+		if b[i].Target == b[j].Target {
+			if b[i].RecordType <= b[j].RecordType {
```
return b[i].RecordType <= b[j].RecordType

and drop the other 2 next return lines

szuecs commented on internal/testutils/endpoint.go:
```diff
@@ -17,14 +17,39 @@ limitations under the License.
+	return false
+}
+
 // SameEndpoint returns true if two endpoint are same
```
endpoint -> endpoints

szuecs commented on internal/testutils/endpoint_test.go:
```diff
@@ -0,0 +1,63 @@
+		{
+			DNSName:    "abc.com",
+			Target:     "1.2.3.4",
+			RecordType: "CNAME",
```
In general CNAME points one DNS name to the other one and not point a name to an IP

ideahitme replied:
yes, but endpoint itself has no knowledge regarding that, maybe makes sense to change that for consistency sake, but where it is relevant I was making sure to be correct with the recordtype

szuecs commented on pkg/apis/externaldns/types.go:
```diff
@@ -71,5 +74,9 @@ func (cfg *Config) ParseFlags(args []string) error {
 	flags.BoolVar(&cfg.DryRun, "dry-run", true, "dry-run mode")
 	flags.BoolVar(&cfg.Debug, "debug", false, "debug mode")
 	flags.BoolVar(&cfg.Version, "version", false, "display the version")
+	flags.StringVar(&cfg.Registry, "registry", "noop", "type of registry for ownership")
```
As long as you have a limited list of providers, I would put it in the docstring. Maybe <noop|txt> is enough to add

szuecs commented on provider/inmemory.go:
```diff
@@ -126,38 +131,50 @@ func (im *InMemoryProvider) validateChangeBatch(zone string, changes *plan.Chang
 		return ErrZoneNotFound
 	}
-	mesh := map[string]bool{}
+	mesh := map[string]map[string]bool{}
```
This type looks too complicated to me. 
All the code below looks very complicated to read, because of that. I think you should create a type and some methods that works on top of it, to have it simpler

ideahitme replied:
totally agree, code looks messy. This is a inmemory dns provider which implements the logic usually sitting on dns provider side, mostly for ease of testing or running external-dns without any real changes but with the ability to track down potential issues, e.g. when two ingress request same hostname. I would probably refactor this code in a separate PR

szuecs replied:
Ok for me

szuecs commented on registry/registry.go:
```diff
@@ -29,3 +29,14 @@ type Registry interface {
 	ApplyChanges(zone string, changes *plan.Changes) error
 }
+
+//TODO(ideahitme): considering moving this to Plan
```
Why don't you move it to endpoint?

You could have:

func (ep Endpoint) Owner() string {
    return ep.Labels[OwnerLabelKey]
}

type Endpoints []*Endpoint
func (eps Endpoints) FilterByOwner(ownerID string) Endpoints {
     filtered := Endpoints{}
     for _, ep := range eps {
             if ep.Owner() == ownerID {
                    filtered = append(filtered, ep)
             }
     }
    return filtered
}

ideahitme replied:
the reason for not defining methods on endpoints is mostly lack of need at the moment. `FilterByOwner` would be a method to be invoked only if ownership is enabled, otherwise it is irrelevant, so IMO it should exist only in the component which requires ownership filtering. 
As for the `SameEndpoints` vs `.Equal` - totally makes sense, however endpoints comparison currently is only needed in tests, hence it is only defined in `internal/testutils`.

szuecs commented on registry/txt.go:
```diff
@@ -0,0 +1,183 @@
+	return &noopNameMapper{}
+}
+
+func (pr *noopNameMapper) toEndpointName(txtDNSName string) string {
```
You don't modify the underlying struct, so there is no need to use a ptr receiver.
use ```func (pr noopNameMapper) toEndpointName(txtDNSName string) string```

szuecs commented on internal/testutils/endpoint.go:
```diff
@@ -17,14 +17,39 @@ limitations under the License.
 // SameEndpoint returns true if two endpoint are same
 // considers example.org. and example.org DNSName/Target as different endpoints
 // TODO:might need reconsideration regarding trailing dot
 func SameEndpoint(a, b *endpoint.Endpoint) bool {
```
I think this should be an Equal method on top of endpoint.Endpoint. If you also do the Owner() method it is also simpler to read

```
func (ep Endpoint) Equal(other *endpoint.Endpoint) bool {
    return ep.DNSName == other.DNSName && ep.Target == other.Target && ep.RecordType == other.RecordType && ep.Owner() == other.Owner()
}
```

szuecs commented on internal/testutils/endpoint.go:
```diff
@@ -37,34 +62,16 @@ func SameEndpoints(a, b []*endpoint.Endpoint) bool {
-				return false
-			}
+	for i := range sa {
+		if !SameEndpoint(sa[i], sb[i]) {
```
this would change to
```if !sa[i].Equal(sb[i]) { return false }```

ideahitme commented:
@szuecs thanks for the review, addressed some of the comments, for the rest gave reasoning to do it later (or not do it).

linki commented on endpoint/endpoint.go:
```diff
@@ -31,15 +31,18 @@ type Endpoint struct {
-		Target:  strings.TrimSuffix(target, "."),
-		Labels:  map[string]string{},
+		DNSName:    strings.TrimSuffix(dnsName, "."),
+		Target:     strings.TrimSuffix(target, "."),
```
I would strongly suggest to use a method for this: `sanitizeHostname` or `removeTrailingDot` as it makes it more obvious to the reader. However, not in this PR.

linki commented on pkg/apis/externaldns/types.go:
```diff
@@ -65,11 +68,15 @@ func (cfg *Config) ParseFlags(args []string) error {
 	flags.BoolVar(&cfg.Version, "version", false, "display the version")
+	flags.StringVar(&cfg.Registry, "registry", "noop", "type of registry for ownership: <noop|txt>")
+	flags.StringVar(&cfg.RecordOwnerID, "record-owner-id", "", "id of the current external dns for labeling owned records")
+	flags.StringVar(&cfg.TXTPrefix, "txt-prefix", "", `prefix of the associated TXT records DNS name; if --txt-prefix=abc,
```
I'd like to prepend the prefix as is. No magic, meaning

| prefix | domain | txt record |
|-------|--------|-----------|
| "" | "example.org" | "example.org" |
| "bar-" | "example.org" | "bar-example.org" |
| "bar." | "example.org" | "bar.example.org" |

linki commented on provider/aws_test.go:
```diff
@@ -507,13 +507,14 @@ func TestAWSSanitizeZone(t *testing.T) {
 func validateEndpoints(t *testing.T, endpoints []*endpoint.Endpoint, expected []*endpoint.Endpoint) {
 	if !testutils.SameEndpoints(endpoints, expected) {
-		t.Errorf("expected %v, got %v", expected, endpoints)
+		spew.Dump(endpoints, expected)
```
drop it

linki commented on provider/aws_test.go:
```diff
@@ -133,24 +134,23 @@ func (r *Route53APIStub) CreateHostedZone(input *route53.CreateHostedZoneInput)
 
 	records := []*endpoint.Endpoint{
-		{DNSName: "create-test.ext-dns-test.teapot.zalan.do", Target: "8.8.8.8"},
+		endpoint.NewEndpoint("create-test.ext-dns-test.teapot.zalan.do", "8.8.8.8", ""),
```
missing type?

linki commented on registry/txt.go:
```diff
@@ -0,0 +1,183 @@
+)
+
+var (
+	txtLabelRegex  = regexp.MustCompile("^heritage=external-dns;record-owner-id=(.+)")
```
to avoid name clashes I would namespace the data that belongs to external-dns with `external-dns/`.

hjacobs commented:
According to @ideahitme:

* small bug to be fixed (updates not working with TXT registry)
* changelog entry

ideahitme commented:
@linki addressed your comments. I would first merge https://github.com/kubernetes-incubator/external-dns/pull/145 and rebase. I will add CHANGELOG in this PR, but detailed documentation for using TXT registry will come as a separate PR

hjacobs commented:
"Go: tip" fails again on Travis (go 1.7 and 1.8 work fine)

hjacobs commented:
Fails:
```
./build/external-dns --provider aws --zone Z16P7IEWFWZ4RB --registry txt --source ingress --source service --record-owner-id hjacobs-test --namespace external-dns --dry-run=false
INFO[0000] targeting cluster at https://kube-aws-test-1.example.org
FATA[0001] InvalidChangeBatch: Invalid Resource Record: FATAL problem: InvalidCharacterString (Value should be enclosed in quotation marks) encountered with 'heritage=external-dns;external-dns/record-owner-id=hjacobs-test'
	status code: 400, request id: 3718635e-1ef7-11e7-a62f-cf235337c794

hjacobs commented:
Basic happy case works now:

```
 ./build/external-dns --provider aws --zone Z16P7IEWFWZ4RB --registry txt --source ingress --source service --record-owner-id hjacobs-test --namespace external-dns --dry-run=false --txt-prefix=_

ideahitme commented:
@hjacobs thanks a lot for help ;) (rebase and test fixes are correct :) )
